hail/src/main/scala/is/hail/expr/ir/agg/Extract.scala

```diff
@@ -391,6 +399,22 @@ object Extract {
           i
         })
         GetTupleElement(result, idx)
+      case x@AggFold(zero, seqOp, combOp, accumName, otherAccumName, isScan) =>
+        val idx = memo.getOrElseUpdate(x, {
+          val i = ab.length
+          val initOpArgs = IndexedSeq(zero)
+          val seqOpArgs = IndexedSeq(seqOp)
+          val op = Fold()
+          val resultEmitType = EmitType(SType.canonical(zero.typ), false) // FIXME: Should be able to be true sometimes.
```

Comment:
This produces a deeply-optional state type. that's going to hurt us for sure.

Can we talk through the issue with the requiredness and refs?

Reply:
I haven't dug into it yet, I can try this afternoon. What I know so far is that sometimes the analysis is deciding that `AggFold` is required, but that at least one of the `Ref`s inside the `seqOp` or `combOp` aren't. The `Requiredness` rule for `AggFold` probably just isn't right.

Reply:
What do you mean by "deeply optional", I thought it was just optional at top level?

Reply:
This piece:
```
SType.canonical(zero.typ)
```

That's currently implemented as 
```
  def canonical(virt: Type): SType = {
    PType.canonical(virt).sType
  }
```

which aliases
```
PType.canonical(t, required=false, innerRequired=false)
```

Reply:
Great, I see. Changed it to:

```
val resultSType = r(x).canonicalPType(x.typ).sType
val resultEmitType = EmitType(resultSType, false)
```

Which should work since it's based on the requiredness analysis.

Reply:
Can we do `r(x).canonicalEmitType`?

Reply:
or does that hit the requiredness mismatch issue above?

Reply:
Currently we can't, working on understanding why.

Reply:
Fixed, see: https://github.com/hail-is/hail/pull/10885/commits/9c940c7afd9c6a2e33c51436eda27bc88b41cda6